### PR TITLE
PR: Send Spyder python path to introspection plugins

### PR DESCRIPTION
**Pull request by rlaverde**
Fixes: #4410

Supersedes #4915
Undo some changes of  #4263

At the moment the spyder path is only used by jedi

### CODE REVIEW

File: spyder/utils/introspection/jedi_plugin.py [hunk 1]
```diff
@@ -30,6 +30,8 @@
 except ImportError:
     jedi = None
 
+JEDI_010 = [int(i) for i in jedi.__version__.split('.')] >= [0, 10, 0]
```
Comment (ccordoba12) [inline spyder/utils/introspection/jedi_plugin.py]: We have `programs.is_module_installed` exactly for this :-)

File: spyder/app/tests/test_mainwindow.py [hunk 2]
```diff
@@ -110,7 +111,7 @@ def main_window(request):
     # (it's faster and less memory consuming not to use it!)
     marker = request.node.get_marker('use_introspection')
     if marker:
-        os.environ['SPY_TEST_USE_INTROSPECTION'] = 'True'
+        os.environ['SPY_TEST_USE_INTROSPECTION'] = 'USE_INTROSPECTION'
```
Comment (ccordoba12) [inline spyder/app/tests/test_mainwindow.py]: Please change this to `True`. That could fix the failing test below.

**Comment by ccordoba12** [comment]
The way to implement this for Rope is the following:

1. The `project` attribute of `RopePlugin` has a `prefs` sub-attribute, as can be seen here:

    https://github.com/spyder-ide/spyder/blob/3.x/spyder/utils/introspection/rope_plugin.py#L74

2. You can set a property in `prefs` called `python_path`, which is a list of additional paths to perform code completion. I think the way to do this is

        self.project.prefs.set('python_path', ['path1', 'path2'])

    but you need to test it to see if it's right.

3. The previous step needs to be done every time we call the `get_completions` method of `RopePlugin`.

File: spyder/app/tests/test_mainwindow.py [hunk 3]
```diff
@@ -110,7 +111,7 @@ def main_window(request):
     # (it's faster and less memory consuming not to use it!)
     marker = request.node.get_marker('use_introspection')
     if marker:
-        os.environ['SPY_TEST_USE_INTROSPECTION'] = 'True'
+        os.environ['SPY_TEST_USE_INTROSPECTION'] = True
```
Comment (ccordoba12) [inline spyder/app/tests/test_mainwindow.py]: This needs to be `'True'`, not `True`.

**Comment by ccordoba12** [comment]
@rlaverde, please remove @dalthviz test because is timing out. Instead, please add tests to `test_rope_plugin.py` and `test_jedi_plugin.py` that get completions of some test module placed in a temp directory.

**Comment by rlaverde** [comment]
>please remove @dalthviz test because is timing out. Instead, please add tests to test_rope_plugin.py and test_jedi_plugin.py that get completions of some test module placed in a temp directory.

I was trying to fix it without success 😞 I'll also try to move it to an editor test, maybe loading the main window is what's causing it to timeout.

**Comment by ccordoba12** [comment]
@rlaverde, I think that adding the test to test_rope_plugin.py and test_jedi_plugin.py is easier and simpler than testing completion in our widgets.

**Comment by rlaverde** [comment]
I think that adding the test to test_rope_plugin.py and test_jedi_plugin.py is easier and simpler than testing completion in our widgets.

Yes, I'll give a last try to the integration test and I'll do that (the test is running pretty well in my local configuration, I don't know way CI is failing 😞 )

**Comment by andfoy** [comment]
@rlaverde Have you tried to debug those tests via SSH on Travis (https://docs.travis-ci.com/user/running-build-in-debug-mode/) and RDP on AppVeyor (https://www.appveyor.com/docs/how-to/rdp-to-build-worker/)?

File: spyder/plugins/tests/test_editor_introspection.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    editor.new(fname="test.py", text="")
+    editor.introspector.set_editor_widget(editor.editorstacks[0])
+
+    yield editor, qtbot
+    # teardown
+    os.environ['SPY_TEST_USE_INTROSPECTION'] = 'False'
+    editor.introspector.plugin_manager.close()
+
+
+@pytest.mark.skipif(True, reason="This test fails too much in the CI :(")
```
Comment (ccordoba12) [inline spyder/plugins/tests/test_editor_introspection.py]: Let's remove this condition and change the one below to

    not JEDI_010 or os.environ.get('CI', False)

so we can run this test locally.

**Comment by ccordoba12** [comment]
@dalthviz, @arman00 and @laikh, please help us to test this PR to see if it fixes the problem with code completion in the Editor when using projects.

Thanks for your help!

**Comment by dalthviz** [comment]
@ccordoba12 It works for me :+1: 🎉 , I tested opening and closing projects and changing the PYTHONPATH.

**Comment by arman00** [comment]
Unfortunately it doesn't work for me. After I cloned from github , I switched to the 3.x branch. Version is 3.2.2.dev0.

It works if I create a new project, but with my existing project which has different modules in different folders, it doesn't work. I add those folders into the PYTHONPATH.

**Comment by dalthviz** [comment]
Hi @arman00 to test this you need to checkout the @rlaverde branch `introspection-python-path`, did you do that?

**Comment by arman00** [comment]
Oh I see... Didn't know that :) 
I will do it as soon as I have the chance again
Thanks

**Comment by dalthviz** [comment]
Just in case, to test this you can clone the [fork](https://github.com/rlaverde/spyder) of @rlaverde and then checkout the corresponding branch.

**Comment by goanpeca** [comment]
You can do

```
$ git clone https://github.com/rlaverde/spyder.git
$ cd spyder
$ git checkout introspection-python-path
```

**Comment by laikh** [comment]
Hi guys,

Finally managed to test your solution and am happy to report that this fixed the issue for me!! :+1: 
I tested it with version 3.2.0 as 3.2.1 seems to be still not available for anaconda.

Thanx a lot, looking forward to the next version!

sincerely,
Michael

**Comment by ccordoba12** [comment]
Great news!! Thanks a lot to you for helping us to test this :-)

**Pull request merged**